# DA3 - Challenges for Week 3


## Assignment 1
Using NetVizz data (either from Amnesty International) or from a page you'd like to investigate, use an NLP function to categorize whether posts and comments contain certain words, and show descriptive statistics and a histogram. Explain in the notebook what the results mean.

**Optional assignment:** How would you consolidate the total number of comments that mention a word, and bring that information at post level? For the data I used in the tutorial, how would we check the percentage of comments that mention Myanmar by post? E.g., post 1 has 10% of the comments mentioning Myanmar, and post 2 has 20% of the comments mentioning Myanmar.

In [2]:
import pandas as pd

I downlaoded the Netvizz data from IKEA facebook page

In [2]:
page_posts = pd.read_table('page_85409903065_2019_02_16_22_16_30_fullstats.tab')
comments = pd.read_table('page_85409903065_2019_02_16_22_16_30_comments.tab')

In [3]:
page_posts

type                     by                        post_id  \
0    status  post_user_85409903065  85409903065_10158432069908066   
1    status  post_user_85409903065  85409903065_10158432007443066   
2    status  post_user_85409903065  85409903065_10158431933553066   
3    status  post_user_85409903065  85409903065_10158431227688066   
4    status  post_user_85409903065  85409903065_10158431227453066   
5    status  post_user_85409903065  85409903065_10158429738513066   
6     photo  post_page_85409903065  85409903065_10158324160238066   
7    status  post_user_85409903065  85409903065_10158429556093066   
8    status  post_user_85409903065  85409903065_10158429538218066   
9    status  post_user_85409903065  85409903065_10158429494363066   
10   status  post_page_85409903065  85409903065_10158429113518066   
11     link  post_page_85409903065  85409903065_10158428971993066   
12   status  post_user_85409903065  85409903065_10158426776613066   
13    photo  post_page_85409903065  85409903065_10158425484923066   
14   status  post_user_85409903065  85409903065_10158425863123066   
15   status  post_user_85409903065  85409903065_10158425568233066   
16   status  post_user_85409903065  85409903065_10158425417013066   
17     link  post_page_85409903065  85409903065_10158422827008066   
18   status  post_user_85409903065  85409903065_10158423751118066   
19   status  post_user_85409903065  85409903065_10158423026858066   
20   status  post_user_85409903065  85409903065_10158422960623066   
21     link  post_page_85409903065  85409903065_10158422462638066   
22   status  post_user_85409903065  85409903065_10158422451448066   
23   status  post_user_85409903065  85409903065_10158422008353066   
24   status  post_user_85409903065  85409903065_10158419793528066   
25   status  post_user_85409903065  85409903065_10158419784113066   
26   status  post_user_85409903065  85409903065_10158419783778066   
27   status  post_user_85409903065  85409903065_10158419783373066   
28   status  post_user_85409903065  85409903065_10158419780843066   
29   status  post_user_85409903065  85409903065_10158419779528066   
..      ...                    ...                            ...   
270    link  post_page_85409903065  85409903065_10158323092993066   
271  status  post_user_85409903065  85409903065_10158321682208066   
272  status  post_user_85409903065  85409903065_10158321660023066   
273  status  post_user_85409903065  85409903065_10158321548253066   
274  status  post_user_85409903065  85409903065_10158321463448066   
275  status  post_user_85409903065  85409903065_10158321434078066   
276    link  post_user_85409903065  85409903065_10158321265943066   
277  status  post_user_85409903065  85409903065_10158321181618066   
278   photo  post_page_85409903065  85409903065_10158310427113066   
279  status  post_user_85409903065  85409903065_10158320656303066   
280  status  post_user_85409903065  85409903065_10158319975568066   
281  status  post_user_85409903065  85409903065_10158319571138066   
282  status  post_user_85409903065  85409903065_10158318878993066   
283   photo  post_page_85409903065  85409903065_10158310925438066   
284  status  post_user_85409903065  85409903065_10158318166058066   
285  status  post_user_85409903065  85409903065_10158316723573066   
286    link  post_user_85409903065  85409903065_10158316274048066   
287  status  post_user_85409903065  85409903065_10158316139018066   
288  status  post_user_85409903065  85409903065_10158316044433066   
289  status  post_user_85409903065  85409903065_10158315603453066   
290  status  post_user_85409903065  85409903065_10158313910498066   
291  status  post_user_85409903065  85409903065_10158313627673066   
292  status  post_user_85409903065  85409903065_10158313603593066   
293  status  post_user_85409903065  85409903065_10158313533308066   
294  status  post_user_85409903065  85409903065_10158313315833066   
295  status  post_user_85409903065  85409903065_10158313216178066   
296  

In [4]:
comments

position                        post_id  \
0         4_0  85409903065_10158431227453066   
1         6_0  85409903065_10158324160238066   
2         6_1  85409903065_10158324160238066   
3         6_2  85409903065_10158324160238066   
4         6_3  85409903065_10158324160238066   
5         6_4  85409903065_10158324160238066   
6         6_5  85409903065_10158324160238066   
7         7_0  85409903065_10158429556093066   
8        10_0  85409903065_10158429113518066   
9        10_1  85409903065_10158429113518066   
10       10_2  85409903065_10158429113518066   
11       10_3  85409903065_10158429113518066   
12       10_4  85409903065_10158429113518066   
13       10_5  85409903065_10158429113518066   
14       10_6  85409903065_10158429113518066   
15       10_7  85409903065_10158429113518066   
16       10_8  85409903065_10158429113518066   
17       10_9  85409903065_10158429113518066   
18      10_10  85409903065_10158429113518066   
19      10_11  85409903065_10158429113518066   
20      10_12  85409903065_10158429113518066   
21      10_13  85409903065_10158429113518066   
22      10_14  85409903065_10158429113518066   
23      10_15  85409903065_10158429113518066   
24       11_0  85409903065_10158428971993066   
25       11_1  85409903065_10158428971993066   
26       11_2  85409903065_10158428971993066   
27       11_3  85409903065_10158428971993066   
28       11_4  85409903065_10158428971993066   
29       11_5  85409903065_10158428971993066   
...       ...                            ...   
1633   297_32  85409903065_10158273635903066   
1634   297_33  85409903065_10158273635903066   
1635   297_34  85409903065_10158273635903066   
1636   297_35  85409903065_10158273635903066   
1637   297_36  85409903065_10158273635903066   
1638   297_37  85409903065_10158273635903066   
1639   297_38  85409903065_10158273635903066   
1640   297_39  85409903065_10158273635903066   
1641   297_40  85409903065_10158273635903066   
1642   297_41  85409903065_10158273635903066   
1643   297_42  85409903065_10158273635903066   
1644   297_43  85409903065_10158273635903066   
1645   297_44  85409903065_10158273635903066   
1646   297_45  85409903065_10158273635903066   
1647   297_46  85409903065_10158273635903066   
1648   297_47  85409903065_10158273635903066   
1649   297_48  85409903065_10158273635903066   
1650   297_49  85409903065_10158273635903066   
1651   297_50  85409903065_10158273635903066   
1652   297_51  85409903065_10158273635903066   
1653   297_52  85409903065_10158273635903066   
1654   297_53  85409903065_10158273635903066   
1655   297_54  85409903065_10158273635903066   
1656   297_55  85409903065_10158273635903066   
1657   297_56  85409903065_10158273635903066   
1658   297_57  85409903065_10158273635903066   
1659   297_58  85409903065_10158273635903066   
1660   297_59  85409903065_10158273635903066   
1661   297_60  85409903065_10158273635903066   
1662   297_61  85409903065_10158273635903066   

                                       post_by  \
0     714d07f0a78b72f9fa577ce8977c972c6b6d02e5   
1     3df3cabd48ea1432e1a7b40c9e6086011323c2f9   
2     3df3cabd48ea1432e1a7b40c9e6086011323c2f9   
3     3df3cabd48ea1432e1a7b40c9e6086011323c2f9   
4     3df3cabd48ea1432e1a7b40c9e6086011323c2f9   
5     3df3cabd48ea1432e1a7b40c9e6086011323c2f9   
6     3df3cabd48ea1432e1a7b40c9e6086011323c2f9   
7     714d07f0a78b72f9fa577ce8977c972c6b6d02e5   
8     3df3cabd48ea1432e1a7b40c9e6086011323c2f9   
9     3df3cabd48ea1432e1a7b40c9e6086011323c2f9   
10    3df3cabd48ea1432e1a7b40c9e6086011323c2f9   
11    3df3cabd48ea1432e1a7b40c9e6086011323c2f9   
12    3df3cabd48ea1432e1a7b40c9e6086011323c2f9   
13    3df3cabd48ea1432e1a7b40c9e6086011323c2f9   
14    3df3cabd48ea1432e1a7b40c9e6086011323c2f9   
15    3df3cabd48ea1432e1a7b40c9e6086011323c2f9   
16    3df3cabd48ea1432e1a7b40c9e6086011323c2f9   
17    3df3cabd48ea1432e1a7b40c9e6086011323c2f9   
18    3df3cabd48ea1432e1a7b40c9e6086011323c2f9   
19    3df3cabd48ea1432e1a7b4

In [18]:
def word_present(text, query):
    import re
    text = str(text).lower()
    query = str(query).lower()
    
    tokens = re.findall(r"[\w']+|[.,!?;$@#]", text)
    if query in tokens:
        return 1
    return 0

In [6]:
page_posts['thank'] = page_posts['post_message'].apply(word_present, args=('thank',))

In [7]:
comments['thank'] = comments['comment_message'].apply(word_present, args=('thank',))

In [8]:
page_posts['thank'].value_counts()

0    287
1     13
Name: thank, dtype: int64

In [9]:
page_posts[(page_posts['thank'] == 1)].describe()

post_published_unix  likes_count_fb  comments_count_fb  \
count         1.300000e+01       13.000000          13.000000   
mean          1.548446e+09       17.307692           3.230769   
std           1.131777e+06       61.803970           6.734907   
min           1.546988e+09        0.000000           0.000000   
25%           1.547504e+09        0.000000           0.000000   
50%           1.548365e+09        0.000000           1.000000   
75%           1.549738e+09        0.000000           3.000000   
max           1.550157e+09      223.000000          25.000000   

       reactions_count_fb  shares_count_fb  engagement_fb  comments_retrieved  \
count           13.000000        13.000000      13.000000           13.000000   
mean            20.307692         0.384615      23.923077            2.538462   
std             72.620457         1.386750      80.561014            5.189980   
min              0.000000         0.000000       0.000000            0.000000   
25%              0.000000         0.000000       1.000000            0.000000   
50%              0.000000         0.000000       1.000000            1.000000   
75%              0.000000         0.000000       3.000000            3.000000   
max            262.000000         5.000000     292.000000           19.000000   

       comments_base  comments_replies  comment_likes_count   rea_LOVE  \
count      13.000000         13.000000            13.000000  13.000000   
mean        1.461538          1.076923             0.692308   2.923077   
std         3.230464          2.139374             2.496151  10.539304   
min         0.000000          0.000000             0.000000   0.000000   
25%         0.000000          0.000000             0.000000   0.000000   
50%         1.000000          0.000000             0.000000   0.000000   
75%         1.000000          1.000000             0.000000   0.000000   
max        12.000000          7.000000             9.000000  38.000000   

       rea_WOW  rea_HAHA    rea_SAD  rea_ANGRY  rea_THANKFUL  thank  
count     13.0      13.0  13.000000       13.0          13.0   13.0  
mean       0.0       0.0   0.076923        0.0           0.0    1.0  
std        0.0       0.0   0.277350        0.0           0.0    0.0  
min        0.0       0.0   0.000000        0.0           0.0    1.0  
25%        0.0       0.0   0.000000        0.0           0.0    1.0  
50%        0.0       0.0   0.000000        0.0           0.0    1.0  
75%        0.0       0.0   0.000000        0.0           0.0    1.0  
max        0.0       0.0   1.000000        0.0           0.0    1.0

In [10]:
comments['thank'].value_counts()

0    1606
1      57
Name: thank, dtype: int64

In [11]:
comments[(comments['thank'] == 1)].describe()

is_reply  comment_like_count  thank
count  57.000000           57.000000   57.0
mean    0.719298            0.877193    1.0
std     0.453336            5.421749    0.0
min     0.000000            0.000000    1.0
25%     0.000000            0.000000    1.0
50%     1.000000            0.000000    1.0
75%     1.000000            0.000000    1.0
max     1.000000           41.000000    1.0

**The results indicated that ‘thank’ appeared 13 times in the page posts, and 57 times in the comments.**

In [1]:
import matplotlib

In [2]:
%matplotlib inline

Question: I cannot produce a histogram, do I have to install anything? Thanks!

## Assignment 2

During Week 2, you should have downloaded/used Twitter data coming from DMI-TCAT, and requested sentiment analysis to us. By now, you should probably already have the results of the sentiment analysis. 

Please combine the Twitter data you have with the sentiment analysis results (merging the dataframes), and explore the data in the notebook. Show some interesting descriptives (potentially using NLP functions to categorize tweets), and interpret the results.

In [3]:
tweets = pd.read_pickle('JingjingWang_export.pkl')
SentimentAnalysis = pd.read_pickle('Jingjing Wang - JingjingWang_EN_completed.pkl')

In [4]:
tweets

id          created_at   from_user_name  \
0       1093857236068302849 2019-02-08 13:01:02         GriffBed   
1       1093857235934068737 2019-02-08 13:01:02         AppLovin   
2       1093857238354198528 2019-02-08 13:01:03   NewsMobilidade   
3       1093857240887541760 2019-02-08 13:01:03     gabriel_irdk   
4       1093857248919633920 2019-02-08 13:01:05      cmccaerdydd   
5       1093857259023683585 2019-02-08 13:01:08            laiki   
6       1093857261146001408 2019-02-08 13:01:08      DKaiserREAL   
7       1093857263373221888 2019-02-08 13:01:09     _busnescymru   
8       1093857263658430464 2019-02-08 13:01:09        AirTunnel   
9       1093857265688473600 2019-02-08 13:01:09        bobbo8181   
10      1093857268020506626 2019-02-08 13:01:10  RomuloLamblet13   
11      1093857270075674624 2019-02-08 13:01:10         ClanMpex   
12      1093857270855811072 2019-02-08 13:01:10    PatriotPennsy   
13      1093857271522750464 2019-02-08 13:01:11     JullyRochaII   
14      1093857272026066944 2019-02-08 13:01:11  blackmediaskate   
15      1093857279177342976 2019-02-08 13:01:12           Be_Sod   
16      1093857280012021760 2019-02-08 13:01:13    flordase_naty   
17      1093857288367099904 2019-02-08 13:01:15  james_a_timmins   
18      1093857299653971968 2019-02-08 13:01:17       techristic   
19      1093857300727648256 2019-02-08 13:01:18       hot21radio   
20      1093857313725779968 2019-02-08 13:01:21           Caequi   
21      1093857319727833088 2019-02-08 13:01:22          HwbNews   
22      1093857327566983168 2019-02-08 13:01:24        aamorizer   
23      1093857332436570113 2019-02-08 13:01:25  CsarAugustoSa11   
24      1093857334315700224 2019-02-08 13:01:26    baletomaanija   
25      1093857335141777408 2019-02-08 13:01:26       aerthonnta   
26      1093857341022371840 2019-02-08 13:01:27       aerthonnta   
27      1093857341215305729 2019-02-08 13:01:27       sahsahesco   
28      1093857341945143296 2019-02-08 13:01:27        qlfoivick   
29      1093857350388236289 2019-02-08 13:01:29    RosendoMateus   
...                     ...                 ...              ...   
237717  1095278486762119168 2019-02-12 11:08:35           vvvmpn   
237718  1095278492793503744 2019-02-12 11:08:36       HYamaguchi   
237719  1095278495347949569 2019-02-12 11:08:37           iRaxy1   
237720  1095278512104267776 2019-02-12 11:08:41     AndreiwKim21   
237721  1095278517120630784 2019-02-12 11:08:42   Thaiis_plicado   
237722  1095278519502950400 2019-02-12 11:08:42        elbritobr   
237723  1095278533780201473 2019-02-12 11:08:46      lismaismail   
237724  1095278535109943297 2019-02-12 11:08:46  nildaalcarinque   
237725  1095278538838675456 2019-02-12 11:08:47     DreemurrKris   
237726  1095278545973243904 2019-02-12 11:08:49        xiuvoces_   
237727  1095278563442470912 2019-02-12 11:08:53     Zete81323679   
237728  1095278570476253185 2019-02-12 11:08:55         yayy_kfc   
237729  1095278570610573312 2019-02-12 11:08:55      Dangonzaga6   
237730  1095278576167993345 2019-02-12 11:08:56     Rajneesh0707   
237731  1095278576008478720 2019-02-12 11:08:56       AdamKasim_   
237732  1095278578088976384 2019-02-12 11:08:56        srmarilia   
237733  1095278578021761024 2019-02-12 11:08:56   Shitao42300484   
237734  1095278580739772417 2019-02-12 11:08:57       Yankaronin   
237735  1095278581402517505 2019-02-12 11:08:57    ihimanshusahu   
237736  1095278582610325505 2019-02-12 11:08:58       love_061_3   
237737  1095278583893889024 2019-02-12 11:08:58          Mobilly   
237738  1095278584795676674 2019-02-12 11:08:58      thamizoando   
237739  1095278586968358912 2019-02-12 11:08:59     NatPennArBed   
237740  1095278595138703360 2019-02-12 11:09:01   leinsecuregirl   
237741  1095278598162993152 2019-02-12 11:09:01     sandraiansen   
237742  1095278606694182918 2019-02-12 11:09:03       LikeImHalf   
237743  1095278608673841152 2019-02-12 11:09:04       nickinglis   
237744  10952

In [5]:
SentimentAnalysis

id  \
1       1093857235934068737   
8       1093857263658430464   
9       1093857265688473600   
11      1093857270075674624   
12      1093857270855811072   
17      1093857288367099904   
18      1093857299653971968   
19      1093857300727648256   
30      1093857353567559684   
37      1093857395762188293   
38      1093857401747300352   
43      1093857426443513856   
47      1093857441240965120   
51      1093857456537653252   
52      1093857461826596870   
57      1093857504331628545   
58      1093857518898610176   
65      1093857556945125376   
66      1093857557041631233   
69      1093857582257786880   
74      1093857620006453253   
79      1093857640827011072   
82      1093857662662598656   
89      1093857704987156480   
94      1093857732522856448   
96      1093857757965496320   
103     1093857785224331266   
105     1093857793193533442   
106     1093857793965080577   
108     1093857810776031233   
...                     ...   
237644  1095278166115987456   
237649  1095278181332840448   
237652  1095278191051190273   
237654  1095278193240608768   
237655  1095278193395814401   
237660  1095278226912403456   
237664  1095278238291685376   
237666  1095278241462583296   
237667  1095278247959396355   
237669  1095278254879981570   
237674  1095278269803450368   
237684  1095278333871280129   
237690  1095278355564412929   
237693  1095278358227836929   
237695  1095278364284387328   
237697  1095278366826090496   
237705  1095278457104285696   
237707  1095278459688009728   
237710  1095278464482099205   
237714  1095278471071199232   
237715  1095278475684896769   
237716  1095278477346029568   
237720  1095278512104267776   
237723  1095278533780201473   
237725  1095278538838675456   
237730  1095278576167993345   
237735  1095278581402517505   
237743  1095278608673841152   
237745  1095278612591259648   
237746  1095278610758361089   

                                                     text negative positive  \
1       .@lewisleong on why the Oculus Go and Oculus Q...       -1        1   
8       AR.Drone Parrot 2.0 Elite Edition in Sand with...       -1        1   
9       RT @heyitsCarolyn: Watch this little snippet f...       -1        2   
11      RT @h1z1hunt: 💰$500 GIVEAWAY!💰\n\n🎁3x AR PATRI...       -1        2   
12      RT @heyitsCarolyn: Watch this little snippet f...       -1        2   
17      RT @Purcelluk: Yesterday morning, our Cambridg...       -2        1   
18      How Lush is using AR to change the way you buy...       -2        1   
19      #NowPlaying : Conversation by AR Williams #lis...       -1        1   
30      RT @claudiomkd: RT Ronald_vanLoon: This app is...       -1        2   
37      RT @heyitsCarolyn: Watch this little snippet f...       -1        2   
38      @maydaymindy9 @PersianKittenz Ar tge end of th...       -1        1   
43      RT @claudiomkd: RT Ronald_vanLoon: This app is...       -1        2   
47      RT @kashthefuturist: This is creepy\n\n#Augmen...       -1        1   
51      AR = Who Pays For Ad Space?: Imagine, if you w...       -2        1   
52      AR (SHARP): latest activity (24h) (server2) #s...       -2        1   
57      RT @_SamKeene: Wow! Thanks for illustrating my...       -1        4   
58      RT @thenextweb: How Lush is using AR to change...       -2        1   
65      How Lush is using AR to change the way you buy...       -2        1   
66      Click to see how easy it is to #augment your b...       -1        1   
69      @KaitMarieox Have a bunch of kids and buy them...       -3        3   
74      @ArchipalagoTV Woho fuuu s3 boliga kramaen!! K...       -3        3   
79      RT @GOGLINJF: Experience #Google Earth in #Vir...       -1        1   
82      @Daisy84x Thanks very much Daisy, that's all r...       -1        3   
89      OTV, ORTEL, MS TELECOM and Calorex holdings, a...       -1        1   
94      RT @kashthefuturist: This is creepy\n\n#Augmen...       -1        1   
96      @lnVayn Matt still , I get it he would b

In [6]:
tweets.columns

Index(['id', 'created_at', 'from_user_name', 'from_user_id', 'from_user_lang',
       'from_user_tweetcount', 'from_user_followercount',
       'from_user_friendcount', 'from_user_listed', 'from_user_realname',
       'from_user_utcoffset', 'from_user_timezone', 'from_user_description',
       'from_user_url', 'from_user_verified', 'from_user_profile_image_url',
       'from_user_created_at', 'from_user_withheld_scope',
       'from_user_favourites_count', 'source', 'location', 'geo_lat',
       'geo_lng', 'text', 'retweet_id', 'retweet_count', 'favorite_count',
       'to_user_id', 'to_user_name', 'in_reply_to_status_id', 'filter_level',
       'lang', 'possibly_sensitive', 'quoted_status_id', 'withheld_copyright',
       'withheld_scope'],
      dtype='object')

In [7]:
SentimentAnalysis.columns

Index(['id', 'text', 'negative', 'positive', 'neutral'], dtype='object')

In [8]:
tweets1 = tweets.merge(SentimentAnalysis, on='id')

In [9]:
tweets1

id          created_at   from_user_name  \
0      1093857235934068737 2019-02-08 13:01:02         AppLovin   
1      1093857263658430464 2019-02-08 13:01:09        AirTunnel   
2      1093857265688473600 2019-02-08 13:01:09        bobbo8181   
3      1093857270075674624 2019-02-08 13:01:10         ClanMpex   
4      1093857270855811072 2019-02-08 13:01:10    PatriotPennsy   
5      1093857288367099904 2019-02-08 13:01:15  james_a_timmins   
6      1093857299653971968 2019-02-08 13:01:17       techristic   
7      1093857300727648256 2019-02-08 13:01:18       hot21radio   
8      1093857353567559684 2019-02-08 13:01:30     InsightBrief   
9      1093857395762188293 2019-02-08 13:01:40    DonnaMWaldrop   
10     1093857401747300352 2019-02-08 13:01:42        gryffon57   
11     1093857426443513856 2019-02-08 13:01:48        Calcaware   
12     1093857441240965120 2019-02-08 13:01:51  kashthefuturist   
13     1093857456537653252 2019-02-08 13:01:55        C3Metrics   
14     1093857461826596870 2019-02-08 13:01:56       SSSWeather   
15     1093857504331628545 2019-02-08 13:02:06        peterxing   
16     1093857518898610176 2019-02-08 13:02:10       startupnws   
17     1093857556945125376 2019-02-08 13:02:19      TechCurated   
18     1093857557041631233 2019-02-08 13:02:19       LogoScanAR   
19     1093857582257786880 2019-02-08 13:02:25         baumanb4   
20     1093857620006453253 2019-02-08 13:02:34       AndrewZigs   
21     1093857640827011072 2019-02-08 13:02:39  BrunaMa65658485   
22     1093857662662598656 2019-02-08 13:02:44      virginmedia   
23     1093857704987156480 2019-02-08 13:02:54     Not_Your_Kiu   
24     1093857732522856448 2019-02-08 13:03:00         AGYG2014   
25     1093857757965496320 2019-02-08 13:03:07      TheVitalize   
26     1093857785224331266 2019-02-08 13:03:13  AmyRose77253769   
27     1093857793193533442 2019-02-08 13:03:15       AndrewZigs   
28     1093857793965080577 2019-02-08 13:03:15         Boopath7   
29     1093857810776031233 2019-02-08 13:03:19          vothelo   
...                    ...                 ...              ...   
71504  1095278166115987456 2019-02-12 11:07:18      lifeofakshy   
71505  1095278181332840448 2019-02-12 11:07:22       anis_nazim   
71506  1095278191051190273 2019-02-12 11:07:24   LiderConectado   
71507  1095278193240608768 2019-02-12 11:07:25       2be_comPlz   
71508  1095278193395814401 2019-02-12 11:07:25    DistressedVol   
71509  1095278226912403456 2019-02-12 11:07:33    littlepotatho   
71510  1095278238291685376 2019-02-12 11:07:35  DorotheaWeyland   
71511  1095278241462583296 2019-02-12 11:07:36      ThinkAspera   
71512  1095278247959396355 2019-02-12 11:07:38        ReginIeIe   
71513  1095278254879981570 2019-02-12 11:07:39  TopCellularDeal   
71514  1095278269803450368 2019-02-12 11:07:43          Aliyy_D   
71515  1095278333871280129 2019-02-12 11:07:58           hemsol   
71516  1095278355564412929 2019-02-12 11:08:03  einfosolutions1   
71517  1095278358227836929 2019-02-12 11:08:04       Jamilumar3   
71518  1095278364284387328 2019-02-12 11:08:05          Tozeboy   
71519  1095278366826090496 2019-02-12 11:08:06     micahshippee   
71520  1095278457104285696 2019-02-12 11:08:28     kamini323436   
71521  1095278459688009728 2019-02-12 11:08:28     TonnieTaller   
71522  1095278464482099205 2019-02-12 11:08:29        Navisjind   
71523  1095278471071199232 2019-02-12 11:08:31        itsyenaac   
71524  1095278475684896769 2019-02-12 11:08:32     Purple_Truth   
71525  1095278477346029568 2019-02-12 11:08:32     TheDonutShop   
71526  1095278512104267776 2019-02-12 11:08:41     AndreiwKim21   
71527  1095278533780201473 2019-02-12 11:08:46      lismaismail   
71528  1095278538838675456 2019-02-12 11:08:47     DreemurrKris   
71529  1095278576167993345 2019-02-12 11:08:56     Rajneesh0707   
71530  1095278581402517505 2019-02-12 11:08:57    ihimanshusahu   
71531  1095278608673841152 2019-02-12 11:09:04       nickinglis   
7153

In [10]:
def categorise_df(row):
    row['sentiment'] = 'neutral'
    if row['neutral'] > 0:
        row['sentiment'] = 'positive'
    if row['neutral'] < 0:
        row['sentiment'] = 'negative'
    return row

In [11]:
tweets1['neutral'] = pd.to_numeric(tweets1['neutral'], errors='coerce')

In [12]:
tweets2 = tweets1.apply(categorise_df, axis=1)

In [13]:
tweets2

id          created_at   from_user_name  \
0      1093857235934068737 2019-02-08 13:01:02         AppLovin   
1      1093857263658430464 2019-02-08 13:01:09        AirTunnel   
2      1093857265688473600 2019-02-08 13:01:09        bobbo8181   
3      1093857270075674624 2019-02-08 13:01:10         ClanMpex   
4      1093857270855811072 2019-02-08 13:01:10    PatriotPennsy   
5      1093857288367099904 2019-02-08 13:01:15  james_a_timmins   
6      1093857299653971968 2019-02-08 13:01:17       techristic   
7      1093857300727648256 2019-02-08 13:01:18       hot21radio   
8      1093857353567559684 2019-02-08 13:01:30     InsightBrief   
9      1093857395762188293 2019-02-08 13:01:40    DonnaMWaldrop   
10     1093857401747300352 2019-02-08 13:01:42        gryffon57   
11     1093857426443513856 2019-02-08 13:01:48        Calcaware   
12     1093857441240965120 2019-02-08 13:01:51  kashthefuturist   
13     1093857456537653252 2019-02-08 13:01:55        C3Metrics   
14     1093857461826596870 2019-02-08 13:01:56       SSSWeather   
15     1093857504331628545 2019-02-08 13:02:06        peterxing   
16     1093857518898610176 2019-02-08 13:02:10       startupnws   
17     1093857556945125376 2019-02-08 13:02:19      TechCurated   
18     1093857557041631233 2019-02-08 13:02:19       LogoScanAR   
19     1093857582257786880 2019-02-08 13:02:25         baumanb4   
20     1093857620006453253 2019-02-08 13:02:34       AndrewZigs   
21     1093857640827011072 2019-02-08 13:02:39  BrunaMa65658485   
22     1093857662662598656 2019-02-08 13:02:44      virginmedia   
23     1093857704987156480 2019-02-08 13:02:54     Not_Your_Kiu   
24     1093857732522856448 2019-02-08 13:03:00         AGYG2014   
25     1093857757965496320 2019-02-08 13:03:07      TheVitalize   
26     1093857785224331266 2019-02-08 13:03:13  AmyRose77253769   
27     1093857793193533442 2019-02-08 13:03:15       AndrewZigs   
28     1093857793965080577 2019-02-08 13:03:15         Boopath7   
29     1093857810776031233 2019-02-08 13:03:19          vothelo   
...                    ...                 ...              ...   
71504  1095278166115987456 2019-02-12 11:07:18      lifeofakshy   
71505  1095278181332840448 2019-02-12 11:07:22       anis_nazim   
71506  1095278191051190273 2019-02-12 11:07:24   LiderConectado   
71507  1095278193240608768 2019-02-12 11:07:25       2be_comPlz   
71508  1095278193395814401 2019-02-12 11:07:25    DistressedVol   
71509  1095278226912403456 2019-02-12 11:07:33    littlepotatho   
71510  1095278238291685376 2019-02-12 11:07:35  DorotheaWeyland   
71511  1095278241462583296 2019-02-12 11:07:36      ThinkAspera   
71512  1095278247959396355 2019-02-12 11:07:38        ReginIeIe   
71513  1095278254879981570 2019-02-12 11:07:39  TopCellularDeal   
71514  1095278269803450368 2019-02-12 11:07:43          Aliyy_D   
71515  1095278333871280129 2019-02-12 11:07:58           hemsol   
71516  1095278355564412929 2019-02-12 11:08:03  einfosolutions1   
71517  1095278358227836929 2019-02-12 11:08:04       Jamilumar3   
71518  1095278364284387328 2019-02-12 11:08:05          Tozeboy   
71519  1095278366826090496 2019-02-12 11:08:06     micahshippee   
71520  1095278457104285696 2019-02-12 11:08:28     kamini323436   
71521  1095278459688009728 2019-02-12 11:08:28     TonnieTaller   
71522  1095278464482099205 2019-02-12 11:08:29        Navisjind   
71523  1095278471071199232 2019-02-12 11:08:31        itsyenaac   
71524  1095278475684896769 2019-02-12 11:08:32     Purple_Truth   
71525  1095278477346029568 2019-02-12 11:08:32     TheDonutShop   
71526  1095278512104267776 2019-02-12 11:08:41     AndreiwKim21   
71527  1095278533780201473 2019-02-12 11:08:46      lismaismail   
71528  1095278538838675456 2019-02-12 11:08:47     DreemurrKris   
71529  1095278576167993345 2019-02-12 11:08:56     Rajneesh0707   
71530  1095278581402517505 2019-02-12 11:08:57    ihimanshusahu   
71531  1095278608673841152 2019-02-12 11:09:04       nickinglis   
7153

In [14]:
tweets2.columns

Index(['id', 'created_at', 'from_user_name', 'from_user_id', 'from_user_lang',
       'from_user_tweetcount', 'from_user_followercount',
       'from_user_friendcount', 'from_user_listed', 'from_user_realname',
       'from_user_utcoffset', 'from_user_timezone', 'from_user_description',
       'from_user_url', 'from_user_verified', 'from_user_profile_image_url',
       'from_user_created_at', 'from_user_withheld_scope',
       'from_user_favourites_count', 'source', 'location', 'geo_lat',
       'geo_lng', 'text_x', 'retweet_id', 'retweet_count', 'favorite_count',
       'to_user_id', 'to_user_name', 'in_reply_to_status_id', 'filter_level',
       'lang', 'possibly_sensitive', 'quoted_status_id', 'withheld_copyright',
       'withheld_scope', 'text_y', 'negative', 'positive', 'neutral',
       'sentiment'],
      dtype='object')

In [16]:
tweets2[['text_x','text_y']]

text_x  \
0      .@lewisleong on why the Oculus Go and Oculus Q...   
1      AR.Drone Parrot 2.0 Elite Edition in Sand with...   
2      RT @heyitsCarolyn: Watch this little snippet f...   
3      RT @h1z1hunt: 💰$500 GIVEAWAY!💰\n\n🎁3x AR PATRI...   
4      RT @heyitsCarolyn: Watch this little snippet f...   
5      RT @Purcelluk: Yesterday morning, our Cambridg...   
6      How Lush is using AR to change the way you buy...   
7      #NowPlaying : Conversation by AR Williams #lis...   
8      RT @claudiomkd: RT Ronald_vanLoon: This app is...   
9      RT @heyitsCarolyn: Watch this little snippet f...   
10     @maydaymindy9 @PersianKittenz Ar tge end of th...   
11     RT @claudiomkd: RT Ronald_vanLoon: This app is...   
12     RT @kashthefuturist: This is creepy\n\n#Augmen...   
13     AR = Who Pays For Ad Space?: Imagine, if you w...   
14     AR (SHARP): latest activity (24h) (server2) #s...   
15     RT @_SamKeene: Wow! Thanks for illustrating my...   
16     RT @thenextweb: How Lush is using AR to change...   
17     How Lush is using AR to change the way you buy...   
18     Click to see how easy it is to #augment your b...   
19     @KaitMarieox Have a bunch of kids and buy them...   
20     @ArchipalagoTV Woho fuuu s3 boliga kramaen!! K...   
21     RT @GOGLINJF: Experience #Google Earth in #Vir...   
22     @Daisy84x Thanks very much Daisy, that's all r...   
23     OTV, ORTEL, MS TELECOM and Calorex holdings, a...   
24     RT @kashthefuturist: This is creepy\n\n#Augmen...   
25     @lnVayn Matt still , I get it he would be amaz...   
26     RT @heyitsCarolyn: Watch this little snippet f...   
27     @ArchipalagoTV Woho fuuu s3 boliga kramaen!! K...   
28     @ARMurugadoss @ivenkatmohan @VishalKOfficial @...   
29     RT @STEPHMHAMILL: My Parents gave me an early ...   
...                                                  ...   
71504  RT @squintneon: Pic 1 - Family of King Abdulla...   
71505  I have been constantly reciting al waqiah, ar-...   
71506  AR Will Spark the Next Big Tech Platform—Call ...   
71507  RT @_______h7: ＊ar magazine Chaeyeon part TRAN...   
71508  Apple's new augmented reality exec shows how i...   
71509  RT @anis_nazim: I have been constantly recitin...   
71510  AR Will Spark the Next Big Tech Platform—Call ...   
71511  AR Will Spark the Next Big Tech Platform—Call ...   
71512  RT @squintneon: Pic 1 - Family of King Abdulla...   
71513  ON SALE! ASUS Zenfone AR 128GB Charcoal Black ...   
71514  RT @dr_mama_one: Sacrificing your own needs to...   
71515  RT @squintneon: Pic 1 - Family of King Abdulla...   
71516  AR Will Spark the Next Big Tech Platform—Call ...   
71517  @sunnex2509 @Oracle_GF @first_of_myKIND @APCNi...   
71518  AR Will Spark the Next Big Tech Platform—Call ...   
71519  Virtual and augmented reality to create real w...   
71520  https://t.co/HSfVKt1iCk AR Will Spark the Next...   
71521  AR Will Spark the Next Big Tech Platform—Call ...   
71522  https://t.co/k8ImOAd5xx AR Will Spark the Next...   
71523  RT @yenapix: izone for ar magazine https://t.c...   
71524  RT @squintneon: Pic 1 - Family of King Abdulla...   
71525  News from Wired: AR Will Spark the Next Big Te...   
71526  RT @_______h7: ＊ar magazine Chaeyeon part TRAN...   
71527  RT @IZONE_DAILY: [📷] 180212 IZ*ONE - ar Magazi...   
71528             @SpeaksInHand // u ar smooth like EGg.   
71529  RT @squintneon: Pic 1 - Family of King Abdulla...   
71530  RT @squintneon: Pic 1 - Family of King Abdulla...   
71531  AR Will Spark the Next Big Tech Platform—Call ...   
71532  RT @kwheeler: AR Will Spark the Next Big Tech ...   
71533  AR Will Spark the Next Big Tech Platform—Call ...   

                                                  text_y  
0      .@lewisleong on why the Oculus Go and Oculus Q...  
1      AR.Drone Parrot 2.0 Elite Edition in Sand with...  
2      RT @heyitsCarolyn: Watch this little snippet f...  
3      RT @h1z1hunt: 💰$500 GIVEAWAY!💰\n\n🎁3x AR PATRI...  
4      RT @heyitsCarolyn: Watch th

Question: I don't understand why after merging, two 'text' columns are created. 

In [19]:
tweets2['Link'] = tweets2['text_y'].apply(word_present, args=('https'or 'http',))

In [20]:
tweets2['Link'].value_counts()

1    41553
0    29981
Name: Link, dtype: int64

In [23]:
withLink = tweets2[tweets2['Link'] == 1]
noLink = tweets2[tweets2['Link'] == 0]

In [22]:
withLink['sentiment'].value_counts()

neutral     22719
positive    12432
negative     6402
Name: sentiment, dtype: int64

In [24]:
noLink['sentiment'].value_counts()

positive    13760
negative     8529
neutral      7692
Name: sentiment, dtype: int64

In [28]:
tweets2['amsterdam'] = tweets2['text_y'].apply(word_present, args=('amsterdam',))

In [29]:
tweets2['amsterdam'].value_counts()

0    71521
1       13
Name: amsterdam, dtype: int64

In [30]:
AMS = tweets2[tweets2['amsterdam'] == 1]

In [31]:
AMS['sentiment'].value_counts()

positive    11
neutral      2
Name: sentiment, dtype: int64

In [37]:
tweets2['IKEA'] = tweets2['text_y'].apply(word_present, args=('ikea',))

In [38]:
IKEA = tweets2[tweets2['IKEA'] == 1]

In [39]:
IKEA['sentiment'].value_counts()

negative    6
neutral     3
positive    1
Name: sentiment, dtype: int64

**The results indicated that:**

1 - There are 41553 posts containing a link, and 29981 without a link.
* Among posts with links, there are 22719 neutral posts, 12432 positive posts, and 6402 negative posts. 
* Among posts without links, there are 7692 neutral posts, 13760 positive posts, and 8529 negative posts. 

2 - There are 13 posts which mentioned "amsterdam"
* Among those posts, 11 are positive,2 are neutral,there is no negative post.  

3 - There are 10 posts which mentioned "ikea"
* Among those posts, 6 are negative, 3 are neutral,1 is negative.  


## Assignment 3
Using Google Analytics (from our course), download data from All Pages (up to 5000) in Site Content, load it into Pandas, and check which are the most popular pages. Use an NLP function to categorize these pages based on their url (maybe some keywords that they contain), and present interesting descriptives about this categorization. The data should be for the past year.



**Note:** Select the Excel format to download from Google Analytics, and when opening the file in Pandas, you need to specify which sheet you want the data from. So something like:

```
data = pd.read_excel('NAMEOFYOURFILE.xlsxl', sheetname='Dataset1')
```

For example:

In [40]:
GAdata = pd.read_excel('Analytics All Web Site Data Pages 20180215-20190215 (1).xlsx', sheet_name='Dataset1')

In [41]:
GAdata

Page  Page Views  \
0                                         /key-trends/         711   
1                                       /key-concepts/         675   
2                                                    /         563   
3             /2018/02/20/big-brother-is-watching-you/         265   
4                                       /social-media/         249   
5    /would-you-trade-your-health-data-for-a-free-a...         121   
6    /2018/03/20/replika-the-robot-that-becomes-you...         113   
7    /your-netflix-recommendations-are-actually-mad...         113   
8                            /boxever-takeoff-for-a-i/         101   
9              /public-relations-a-digital-revolution/          79   
10   /2018/02/20/how-big-data-polarizes-your-politi...          74   
11                         /big-public-data-relations/          74   
12           /public-relations-a-digital-revolution-2/          69   
13                     /key-concepts/social-listening/          59   
14   /would-you-trade-your-health-data-for-a-free-a...          59   
15    /digital-analytics-and-political-communications/          58   
16   /health-communication-and-big-data-a-win-win-s...          57   
17   /your-netflix-recommendations-are-actually-mad...          57   
18                          /programmatic-advertising/          56   
19               /how-big-data-reinforces-advertising/          50   
20   /2018/03/12/health-communication-and-big-data-...          45   
21                       /big-public-data-relations-2/          44   
22                     /key-concepts/machine-learning/          44   
23                         /our-social-media-presence/          43   
24                                       /page-1-gdpr/          40   
25                              /internet-of-things-2/          39   
26               /2018/03/13/the-rise-of-ehealth-data/          38   
27                       /key-concepts/microtargeting/          35   
28   /2018/03/02/netflix-knows-you-better-than-your...          34   
29                                /amazon-go-ai-group/          34   
..                                                 ...         ...   
193                        /blog-spotify/?preview=true           1   
194                         /boxever-takeoff-for-a-i2/           1   
195  /key-concepts/?preview_id=399&preview_nonce=68...           1   
196        /key-concepts/location-based-advertising-1/           1   
197                          /key-trends/?preview=true           1   
198  /page-2-gdpr/?preview_id=787&preview_nonce=6fc...           1   
199  /page/3/?fbclid=IwAR0fxTG6gKn784tDF5Ne4oFHRGTx...           1   
200  /page/3/?fbclid=IwAR0XlzEPtjNhu5qFFIN9YToD9i0B...           1   
201                                           /page/7/           1   
202          /public-relations-a-digital-revolution/2/           1   
203                /public-relations…gital-revolution/           1   
204  /serie/outlander/temporada-4/episodio-7?fbclid...           1   
205                                  /slippery-slope-1           1   
206                                   /slippery-slope1           1   
207                                   /social-media-2/           1   
208                            /social-media-presence/           1   
209                /the-rise-of-data…f-digiceutical-2/           1   
210  /the-rise-of-robotics-and-ai-2/?preview_id=940...           1   
211                                  /topic-example-2/           1   
212                                     /topic-name-a/           1   
213                                     /trial-page-1/           1   
214  /trial-page-1/?preview_id=726&preview_nonce=28...           1   
215                                     /trial-page-2/           1   
216                                         /wordpress           1   
217                  /wp-admin/edit.php?post_type=page           1   
218              /wp-admin/post-new.php?post_type=page      

In [42]:
def traffic_group (x):
    if x > 100:
        return 'high'
    if x <= 100:
        return 'low'

In [43]:
GAdata['traffic_group'] = GAdata['Page Views'].apply(traffic_group)

In [44]:
High_Traffic = GAdata[GAdata['traffic_group'] == 'high']

In [45]:
High_Traffic[['Page','Page Views','Avg. Time on Page','Bounce Rate']]

Page  Page Views  \
0                                         /key-trends/         711   
1                                       /key-concepts/         675   
2                                                    /         563   
3             /2018/02/20/big-brother-is-watching-you/         265   
4                                       /social-media/         249   
5    /would-you-trade-your-health-data-for-a-free-a...         121   
6    /2018/03/20/replika-the-robot-that-becomes-you...         113   
7    /your-netflix-recommendations-are-actually-mad...         113   
8                            /boxever-takeoff-for-a-i/         101   
222                                                NaN        5354   

     Avg. Time on Page  Bounce Rate  
0            35.489028     0.019108  
1            43.290378     0.176471  
2           103.725714     0.262411  
3           360.611111     0.756522  
4            52.661616     0.269841  
5            59.158416     0.000000  
6           290.227273     0.468750  
7            74.214286     0.181818  
8            28.318182     0.142857  
222          73.281954     0.391616

**The 8 URLs above are the most popular pages which gained traffic over 100.**

In [46]:
GAdata['Data'] = GAdata['Page'].apply(word_present, args=('data',))

In [49]:
Data_related = GAdata[GAdata['Data'] == 1]
non_Data = GAdata[GAdata['Data'] == 0]

In [50]:
Data_related.describe()

Page Views  Unique Page Views  Avg. Time on Page  Entrances  \
count   39.000000          39.000000          39.000000  39.000000   
mean    22.025641          15.179487         137.384594   4.923077   
std     26.929232          17.374924         189.425794  10.451353   
min      1.000000           1.000000           0.000000   0.000000   
25%      1.500000           1.000000          21.000000   0.500000   
50%     11.000000           9.000000          68.833333   1.000000   
75%     33.500000          22.500000         143.793860   5.000000   
max    121.000000          69.000000         735.500000  61.000000   

       Bounce Rate     % Exit  Page Value  Data  
count    39.000000  39.000000        39.0  39.0  
mean      0.230574   0.381243         0.0   1.0  
std       0.337836   0.325619         0.0   0.0  
min       0.000000   0.000000         0.0   1.0  
25%       0.000000   0.139141         0.0   1.0  
50%       0.000000   0.333333         0.0   1.0  
75%       0.468750   0.512712         0.0   1.0  
max       1.000000   1.000000         0.0   1.0

In [51]:
non_Data.describe()

Page Views  Unique Page Views  Avg. Time on Page    Entrances  \
count   184.000000         184.000000         184.000000   184.000000   
mean     53.527174          34.239130          74.381503    17.097826   
std     402.433403         257.415607         119.739959   127.106088   
min       1.000000           1.000000           0.000000     0.000000   
25%       1.000000           1.000000           3.375000     0.000000   
50%       3.000000           2.000000          34.090909     1.000000   
75%      17.000000          12.250000          98.416667     3.000000   
max    5354.000000        3446.000000         847.000000  1669.000000   

       Bounce Rate      % Exit  Page Value   Data  
count   184.000000  184.000000       184.0  184.0  
mean      0.256753    0.414687         0.0    0.0  
std       0.367012    0.391336         0.0    0.0  
min       0.000000    0.000000         0.0    0.0  
25%       0.000000    0.000000         0.0    0.0  
50%       0.000000    0.333333         0.0    0.0  
75%       0.500000    0.784661         0.0    0.0  
max       1.000000    1.000000         0.0    0.0

**The results indicated that 39 pages are data-related, and 184 pages are non-data-related (simply by the url name).**
* The average traffic for data-related pages is 22 while that for non-data related pages is 54.
* The average bounce rate for data-related pages is 23.1% while that for non-data related pages is 25.7%, both performing well.